forestdb Transaction Error on Linux

During `ReplicationInternalTest. testOneShotPullReplBulkGetWith404` on Linux, null transaction error occurs.


> [ERROR] [system.err] java: /home/hideki/github/couchbase-lite-java/libraries/couchbase-lite-java-forestdb/vendor/cbforest/C/c4Database.cc:103: forestdb::Transaction *c4Database::transaction(): Assertion `_transaction' failed.


Log: https://gist.github.com/hideki/1fd28b3c1db29a047596


According to jens, this could be caused by accessing forestdb from multiple threads simultaneously.
> Is it possible you're accidentally calling the same C4Database from multiple threads? So far that's the only way I can see for this to happen.

--- a/src/main/java/com/couchbase/lite/replicator/PullerInternal.java
+++ b/src/main/java/com/couchbase/lite/replicator/PullerInternal.java
@@ -376,14 +376,11 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
      */
     private boolean shouldRetryDownload(String docId) {
         Map<String, Object> localDoc = getLocalDatabase().getExistingLocalDocument(docId);
+
         if (localDoc == null) {
-            Map<String, Object> props = new HashMap<String, Object>();
+            final Map<String, Object> props = new HashMap<String, Object>();
             props.put("retryCount", 1L);
-            try {
-                getLocalDatabase().putLocalDocument(docId, props);
-            } catch (CouchbaseLiteException e) {
-                Log.w(TAG, "Failed to store retryCount value for docId: " + docId, e);
-            }
+            putLocalDocument(docId, props);
             return true;
         }
 
@@ -394,20 +391,34 @@ public class PullerInternal extends ReplicationInternal implements ChangeTracker
         }
 
         localDoc.put("retryCount", retryCount + 1);
-        try {
-            getLocalDatabase().putLocalDocument(docId, localDoc);
-        } catch (CouchbaseLiteException e) {
-            Log.w(TAG, "Failed to store retryCount value for docId: " + docId, e);
-        }
+        putLocalDocument(docId, localDoc);
         return true;
     }
 
-    private void pruneFailedDownload(String docId) {
-        try {
-            getLocalDatabase().deleteLocalDocument(docId);
-        } catch (CouchbaseLiteException e) {
-            Log.w(TAG, "Failed to delete local document: " + docId, e);
-        }
+    private void putLocalDocument(final String docId, final Map<String, Object> localDoc) {
+        workExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    getLocalDatabase().putLocalDocument(docId, localDoc);
+                } catch (CouchbaseLiteException e) {
+                    Log.w(TAG, "Failed to store retryCount value for docId: " + docId, e);
+                }
+            }
+        });
+    }
+
+    private void pruneFailedDownload(final String docId) {
+        workExecutor.submit(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    getLocalDatabase().deleteLocalDocument(docId);
+                } catch (CouchbaseLiteException e) {
+                    Log.w(TAG, "Failed to delete local document: " + docId, e);
+                }
+            }
+        });
     }
 
     // This invokes the tranformation block if one is installed and queues the resulting CBL_Revision

